### PR TITLE
Use WaitQueue to implement waiting in FutexWaiter

### PR DESCRIPTION
**Pull request by sdww0**
Related issue: #1048

### CODE REVIEW

File: kernel/aster-nix/src/process/posix_thread/futex.rs [hunk 1]
```diff
@@ -423,17 +431,31 @@ impl FutexWaiter {
...
         if current_thread.tid() != self.tid {
             return;
         }
         self.is_woken.store(false, Ordering::SeqCst);
-        while !self.is_woken() {
-            // debug!("futex is wait for waken, tid = {}", self.tid);
-            Thread::yield_now();
+
+        if let Some(timeout) = timeout {
+            let wait_queue = WaitQueue::new();
```
Comment (StevenJiang1110) [inline kernel/aster-nix/src/process/posix_thread/futex.rs:442]: The `wait_queue` is created as a local variable, so the `wake` method of this `wait_queue` can never be called. 

I suggest putting the `wait_queue` in `FutexWaiter`, and the method `FutexWaiter::wake` should also wake up the `wait_queue`

File: kernel/aster-nix/src/process/posix_thread/futex.rs [hunk 2]
```diff
@@ -423,17 +431,31 @@ impl FutexWaiter {
...
+                    Some(())
+                } else {
+                    None
+                }
+            };
+
+            wait_queue.wait_until_or_timeout(wake_cond, &timeout.duration);
+        } else {
+            // The PtyTest.TermiosPollReplica test will stuck if we use WaitQueue.
+            while !self.is_woken() {
```
Comment (StevenJiang1110) [inline kernel/aster-nix/src/process/posix_thread/futex.rs:454]: Could you please try again whether the branch can be removed when doing modifications according to the former comment?

File: kernel/aster-nix/src/process/posix_thread/futex.rs [hunk 3]
```diff
@@ -423,17 +431,31 @@ impl FutexWaiter {
...
+                // debug!("futex is wait for waken, tid = {}", self.tid);
+                Thread::yield_now();
+            }
         }
-        // debug!("futex is waken, tid = {}", self.tid);
```
Comment (StevenJiang1110) [inline kernel/aster-nix/src/process/posix_thread/futex.rs:436]: Could you please remove this unused documentations ( alghouth they are not introduced by this PR)?

**Review by StevenJiang1110 — COMMENTED** [review]
This PR is accepable as a temporary fix! I only have some minor comments.

**Review by StevenJiang1110 — APPROVED** [review]
This PR now looks good to me!

**Comment by tatetian** [comment]
Could you explain the rationale of adding vdso as a Git submodule? We have none of them until now. So it is quite a radical change to the codebase.

**Comment by sdww0** [comment]
> Could you explain the rationale of adding vdso as a Git submodule? We have none of them until now. So it is quite a radical change to the codebase.

I accidentally added the vdso as a Git submodule when switching from the benchmark branch to the dev branch. I have removed it in the latest commit.

**Review by tatetian — COMMENTED** [review]
Thanks for adding timeout support for futex. And you know what, @cqs21 measured the thread benchmark with this PR and witnessed a 5X improvement in latency. Good job!

While adding a new feature is better than no feature, I don't want to leave obvious bugs or limitations in our implementation of the feature (let's say, futex timeout) and let future "us" to spend extra time on "rediscovering" these bugs when applications work unexpected.

In this PR, the obvious bug/limitation is the timeout is assumed to a relative value and be based on the monotonic clock. But in reality, the timeout may be either _absolute_ or _relative_ and may be based on the _monotonic_ or _realtime_ clock.

> FUTEX_WAIT
> Note: for FUTEX_WAIT, timeout is interpreted as a relative value.
>
> FUTEX_WAIT_BITSET
> If timeout is not NULL, the structure it points to specifies an absolute timeout for the wait operation.
>
> FUTEX_CLOCK_REALTIME
> If this option is set, the kernel measures the timeout against the CLOCK_REALTIME clock. If this option is not set, the kernel measures the timeout against the CLOCK_MONOTONIC clock.

**Comment by sdww0** [comment]
Since I'm not familiar with the `aster-nix` and syscall manual, I think I should focus on benchmarking to identify the gaps in each benchmark. This PR won't include the implementation of the timeout in futex wait. Still, it will use `WaitQueue` to implement the wait operation for FutexWaiter, which could be the reason for the 5X improvement in latency.

**Comment by tatetian** [comment]
> > Thanks for adding timeout support for futex. And you know what, @cqs21 measured the thread benchmark with this PR and witnessed a 5X improvement in latency. Good job!
> > While adding a new feature is better than no feature, I don't want to leave obvious bugs or limitations in our implementation of the feature (let's say, futex timeout) and let future "us" to spend extra time on "rediscovering" these bugs when applications work unexpected.
> > In this PR, the obvious bug/limitation is the timeout is assumed to a relative value and be based on the monotonic clock. But in reality, the timeout may be either _absolute_ or _relative_ and may be based on the _monotonic_ or _realtime_ clock.
> > > FUTEX_WAIT
> > > Note: for FUTEX_WAIT, timeout is interpreted as a relative value.
> > > FUTEX_WAIT_BITSET
> > > If timeout is not NULL, the structure it points to specifies an absolute timeout for the wait operation.
> > > FUTEX_CLOCK_REALTIME
> > > If this option is set, the kernel measures the timeout against the CLOCK_REALTIME clock. If this option is not set, the kernel measures the timeout against the CLOCK_MONOTONIC clock.
> 
> Since I'm not familiar with the `aster-nix` and syscall manual, I think I should focus on benchmarking to identify the gaps in each benchmark. This PR won't include the implementation of the timeout in futex wait. Still, it will use `WaitQueue` to implement the wait operation in FutexWaiter, which could be the reason for the 5X improvement in latency.

Fair enough.

> In this PR, the obvious bug/limitation is the timeout is assumed to a relative value and be based on the monotonic clock. But in reality, the timeout may be either absolute or relative and may be based on the monotonic or realtime clock.

@StevenJiang1110 Could you take the job of submitting a follow-up PR to make the futex-timeout feature production ready?

**Pull request merged**